src/update-course-links.js

```diff
@@ -0,0 +1,199 @@
+'use strict';
+
+const parser = require('parse5');
+const fs = require('fs');
+const path = require('path');
+
+const { LPVersion } = require('./constants');
+
+module.exports = class CourseLinkUpdater {
+	constructor(
+		{
+			contentDirectory,
+			manifestPath,
+			isDryRun
+		},
+		valence,
+		fetch = require('node-fetch'),
+		ModuleProcessor = require('./utility/module-processor')
+	) {
+		this._fetch = fetch;
+		this._valence = valence;
+		this._contentPath = contentDirectory;
+		this._manifestPath = manifestPath;
+		this._dryRun = isDryRun;
+
+		this._markdownRegex = /.md$/i;
+		this._htmlRegex = /.html$/i;
+
+		this._moduleProcessor = new ModuleProcessor({ contentPath: contentDirectory, isDryRun }, valence);
+	}
+
+	async updateLinks(
+		instanceUrl,
+		orgUnitId,
+		structure
+	) {
+		const orgUnit = await this._getOrgUnit(instanceUrl, orgUnitId);
+
+		const manifest = await this._getManifest();
+		for (const module of manifest.modules) {
+			/* eslint-disable no-await-in-loop */
+			await this._processModule(orgUnit, module, structure, manifest);
+
+			await this._moduleProcessor.processModule(instanceUrl, orgUnit, module);
+			/* eslint-enable no-await-in-loop */
+		}
+	}
+
+	async _processModule(orgUnit, module, structure, manifest) {
+		const fileName = module.descriptionFileName.replace(this._markdownRegex, '.html');
+		const file = await fs.promises.readFile(`${this._contentPath}/${fileName}`);
+		const directory = path.dirname(fileName);
+
+		console.log(`Processing file: '${fileName}'`);
+
+		const document = parser.parse(file.toString('utf-8'));
+		const links = this._getAllLinks(document);
+
+		for (const link of links) {
+			const href = link.attrs.find(x => x.name === 'href');
+			if (!href || href.value.includes('://') || href.value.includes('/d2l')) {
+				continue;
+			}
+
+			// Found a relative link
+			const hrefMd = href.value.replace(this._htmlRegex, '.md');
+			const mdPath = `${directory}/${hrefMd}`;
+			const entry = this._findInManifest(x => x.fileName === mdPath || x.descriptionFileName === mdPath, manifest);
+
+			if (entry.type === 'module') {
+				const foundModule = Array.isArray(structure) && structure.find(m => m.Type === 0 && m.Title === entry.title);
+
+				const newHref = `/d2l/le/lessons/${orgUnit.Identifier}/units/${foundModule.Id}`;
```

Comment:
This makes an educated assumption that we will continue to use the lessons UI.
Otherwise linking becomes more *fun* as modules have no official route. (There is a querystring hack but it's not ideal)